runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -446,6 +577,173 @@ public void restoreState(StreamTaskState state) throws Exception {
+      dataOut.writeInt(timer.f0.limit());
+      dataOut.write(timer.f0.array(), 0, timer.f0.limit());
+      timerCoder.encode(timer.f1, dataOut, Coder.Context.NESTED);
+    }
+  }
+
+  private void registerEventTimeTimer(TimerInternals.TimerData timer) {
+    Tuple2<ByteBuffer, TimerInternals.TimerData> keyedTimer =
+        new Tuple2<>((ByteBuffer) getStateBackend().getCurrentKey(), timer);
+    if (watermarkTimers.add(keyedTimer)) {
+      watermarkTimersQueue.add(keyedTimer);
+    }
+  }
+
+  private void deleteEventTimeTimer(TimerInternals.TimerData timer) {
+    Tuple2<ByteBuffer, TimerInternals.TimerData> keyedTimer =
+        new Tuple2<>((ByteBuffer) getStateBackend().getCurrentKey(), timer);
+    if (watermarkTimers.remove(keyedTimer)) {
+      watermarkTimersQueue.remove(keyedTimer);
+    }
+  }
+
+  private void registerProcessingTimeTimer(TimerInternals.TimerData timer) {
+    Tuple2<ByteBuffer, TimerInternals.TimerData> keyedTimer =
+        new Tuple2<>((ByteBuffer) getStateBackend().getCurrentKey(), timer);
+    if (processingTimeTimers.add(keyedTimer)) {
+      processingTimeTimersQueue.add(keyedTimer);
+
+      // If this is the first timer added for this timestamp register a timer Task
+      if (processingTimeTimerTimestamps.add(timer.getTimestamp().getMillis(), 1) == 0) {
+        ScheduledFuture<?> scheduledFuture = registerTimer(timer.getTimestamp().getMillis(), this);
+        processingTimeTimerFutures.put(timer.getTimestamp().getMillis(), scheduledFuture);
+      }
+    }
+  }
+
+  private void deleteProcessingTimeTimer(TimerInternals.TimerData timer) {
+    Tuple2<ByteBuffer, TimerInternals.TimerData> keyedTimer =
+        new Tuple2<>((ByteBuffer) getStateBackend().getCurrentKey(), timer);
+    if (processingTimeTimers.remove(keyedTimer)) {
+      processingTimeTimersQueue.remove(keyedTimer);
+
+      // If there are no timers left for this timestamp, remove it from queue and cancel the
+      // timer Task
+      if (processingTimeTimerTimestamps.remove(timer.getTimestamp().getMillis(), 1) == 1) {
+        ScheduledFuture<?> triggerTaskFuture =
+            processingTimeTimerFutures.remove(timer.getTimestamp().getMillis());
+        if (triggerTaskFuture != null && !triggerTaskFuture.isDone()) {
+          triggerTaskFuture.cancel(false);
+        }
+      }
+
+    }
+  }
+
+  @Override
+  public void trigger(long time) throws Exception {
+
+    if (keyCoder != null) {
+      //Remove information about the triggering task
+      processingTimeTimerFutures.remove(time);
+      processingTimeTimerTimestamps.setCount(time, 0);
+
+      consumeTimers(processingTimeTimersQueue, processingTimeTimers, time);
+
+    } else {
+      throw new RuntimeException("unexpected trigger: " + time);
+    }
+
+  }
+
+  protected void fireTimer(Tuple2<ByteBuffer, TimerInternals.TimerData> timer) {
+    TimerInternals.TimerData timerData = timer.f1;
+    pushbackDoFnRunner.onTimer(timerData.getTimerId(), GlobalWindow.INSTANCE,
```

Comment:
@kennknowles I use GlobalWindow refer to DirectRunner(ExecutorServiceParallelExecutor.fireTimers). Should I preserve BoundedWindow when user setTimer? I do not quite understand its meaning.